src/typescript/frontend/src/lib/store/name-store.ts

```diff
@@ -0,0 +1,173 @@
+
+/**
+ * Stores promises to requests.
+ *
+ * This is used to avoid duplicating requests when two sources need the name of one address at the same time.
+ */
+const ansPromiseMap: ANSPromiseMap = new Map();
+
+function getRequestURL(address: string, isPrimary: boolean) {
+  return isPrimary
+    ? `https://www.aptosnames.com/api/${APTOS_NETWORK}/v1/primary-name/${address}`
+    : `https://www.aptosnames.com/api/${APTOS_NETWORK}/v1/name/${address}`;
+}
+
+/**
+ * Saves the ANSMap to localStorage.
+ *
+ * Removes outdated values by the way.
+ */
+export function saveANSMap(map: ANSMap) {
+  const now = new Date().getTime();
+  const notExpired = (entry: [string, ANSValue]) => entry[1].expiry > now;
+
+  const entries = Array.from(map.entries())
+    .filter(notExpired)
+    .sort((a, b) => b[1].expiry - a[1].expiry)
+    .slice(0, MAX_ITEMS_IN_CACHE);
+
+  const serializedMap = JSON.stringify(entries);
+
+  localStorage.setItem(LOCALSTORAGE_ANS_KEY, serializedMap);
+}
+
+/** Loads the ANSMap from localStorage. */
+export function loadANSMap(): ANSMap {
+  const map = localStorage.getItem(LOCALSTORAGE_ANS_KEY);
+  if (map) {
+    return new Map(JSON.parse(map));
+  } else {
+    return new Map();
+  }
+}
+
+/**
+ * Get the name of an address.
+ *
+ * Returns null if the address has no registered name.
+ *
+ * Do not expose this function to avoid using it in a for loop.
+ */
+async function fetchANSName(address: string): Promise<string | null> {
+  const primaryNameUrl = getRequestURL(address, true);
+
+  try {
+    const { name: primaryName } = await fetch(primaryNameUrl).then((res) => res.json());
+
+    if (primaryName) {
+      return primaryName;
+    } else {
+      const nameUrl = getRequestURL(address, false);
+
+      if (!nameUrl) {
+        return null;
+      }
+
+      const { name } = await fetch(nameUrl).then((res) => res.json());
+      if (name) {
+        return name;
+      }
+      return null;
+    }
+  } catch (_) {
+    return null;
+  }
+}
+
+export const initializeEventStore = (): NameState => {
+  if (typeof localStorage !== "undefined") {
+    return { names: loadANSMap() };
+  } else {
+    return { names: new Map() };
+  }
+};
+
+export const defaultState: NameState = initializeEventStore();
+
+type Set = (
+  nextStateOrUpdater: NameStore | Partial<NameStore> | ((state: WritableDraft<NameStore>) => void),
+  shouldReplace?: boolean | undefined
+) => void;
+type Get = () => NameStore;
+
+const getNameMap = (set: Set, get: Get, addresses: string[]): ANSMap => {
+  const map = get().names;
+  const uncachedNames: string[] = [];
+  for (const address of addresses) {
+    if (!map.has(address)) {
+      uncachedNames.push(address);
+    }
+  }
+  if (uncachedNames.length > 0) {
+    for (const address of uncachedNames) {
+      if (!ansPromiseMap.has(address)) {
+        const x = fetchANSName(address)
+          .then((res) => ({
+            name: res,
+            isNull: res === null,
```

Comment:
Don't think you use this, yea? Did you mean to

Reply:
Yes I do use it, line 153. It is stored in the promise map.

Reply:
`isNull`? I didn't mean the value `x`, just the `isNull` field, sorry

Reply:
Feel free to resolve, not a big deal either way